### PR TITLE
Image map support

### PR DESCRIPTION
**Pull request by simonferquel**
~~This add support for pushing / pull from and to an OCIIndex/DockerManifestList~~ deferred to a followup PR
Adds support for creating image maps, and do image remapping on install (for scenarios like applevel promotion and airgapped scenarios with DTR).

### CODE REVIEW

**Comment by simonferquel** [comment]
@silvin-lubecki @rumpl I rebased this one, and stripped the cnab-to-oci integration from it (for now - I backid it up in separate branch). can you PTAL ?

File: cmd/cnab-run/install.go [hunk 1]
```diff
@@ -13,6 +16,18 @@ import (
 	"github.com/spf13/pflag"
 )
 
+func getBundleImageMap() (map[string]bundle.Image, error) {
```
Comment (silvin-lubecki) [inline cmd/cnab-run/install.go]: nit: can you move this function under the `install` function?

File: cmd/docker-app/render.go [hunk 2]
```diff
@@ -39,7 +40,7 @@ func renderCmd(dockerCli command.Cli) *cobra.Command {
 			}
 			defer app.Cleanup()
 			d := cliopts.ConvertKVStringsToMap(renderEnv)
-			rendered, err := render.Render(app, d)
+			rendered, err := render.Render(app, d, map[string]bundle.Image{})
```
Comment (silvin-lubecki) [inline cmd/docker-app/render.go]: Use a nil instead?

File: examples_test.go [hunk 3]
```diff
@@ -18,12 +19,12 @@ func Example() {
...
 	app, err := loader.LoadFromSingleFile("myApp", f)
 	if err != nil {
-		panic("cannot load application")
+		panic("cannot load application: " + err.Error())
 	}
 	// Render the app to a composefile format, using some user provided parameters
 	c, err := render.Render(app, map[string]string{
 		"text": "hello examples!",
-	})
+	}, map[string]bundle.Image{})
```
Comment (silvin-lubecki) [inline examples_test.go]: Use nil instead?

File: cmd/docker-app/image-add.go [hunk 4]
```diff
@@ -37,7 +38,7 @@ subdirectory.`,
 			}
 			defer app.Cleanup()
 			d := cliopts.ConvertKVStringsToMap(imageAddEnv)
-			config, err := render.Render(app, d)
+			config, err := render.Render(app, d, map[string]bundle.Image{})
```
Comment (simonferquel) [inline cmd/docker-app/image-add.go]: use nil instead of empty map

File: internal/packager/cnab.go [hunk 5]
```diff
@@ -58,6 +59,23 @@ func ToCNAB(app *types.App, invocationImageName string) *bundle.Bundle {
 	if app.Metadata().Namespace != "" {
 		name = fmt.Sprintf("%s/%s", app.Metadata().Namespace, name)
 	}
+
+	_, images, err := compose.Load(app.Composes(), func(v string) (string, error) { return v, nil })
```
Comment (silvin-lubecki) [inline internal/packager/cnab.go]: Extract and move this to a subfunction?

File: render/render_test.go [hunk 6]
```diff
@@ -32,7 +33,7 @@ func TestRenderMissingValue(t *testing.T) {
 	finalEnv := map[string]string{
 		"imageName": "foo",
 	}
-	_, err := render(configFiles, finalEnv)
+	_, err := render(configFiles, finalEnv, map[string]bundle.Image{})
```
Comment (silvin-lubecki) [inline render/render_test.go]: Use nil instead?

File: render/render_test.go [hunk 7]
```diff
@@ -55,7 +56,7 @@ func TestRender(t *testing.T) {
 		"version": "latest",
 		"foo.bar": "baz",
 	}
-	c, err := render(configFiles, finalEnv)
+	c, err := render(configFiles, finalEnv, map[string]bundle.Image{})
```
Comment (silvin-lubecki) [inline render/render_test.go]: Use nil instead?

File: render/render_test.go [hunk 8]
```diff
@@ -80,7 +81,7 @@ func TestRenderEnabledFalse(t *testing.T) {
 		}
 		c, err := render(configs, map[string]string{
 			"myapp.debug": "true",
-		})
+		}, map[string]bundle.Image{})
```
Comment (silvin-lubecki) [inline render/render_test.go]: Use nil instead?

File: cmd/cnab-run/install.go [hunk 9]
```diff
@@ -53,3 +60,15 @@ func getFlagset(orchestrator command.Orchestrator) *pflag.FlagSet {
 	}
 	return result
 }
+
+func getBundleImageMap() (map[string]bundle.Image, error) {
+	mapJSON, err := ioutil.ReadFile("/cnab/app/image-map.json")
```
Comment (silvin-lubecki) [inline cmd/cnab-run/install.go]: Extract the file name as a documented const?

File: render/render.go [hunk 10]
```diff
@@ -81,6 +82,12 @@ func render(configFiles []composetypes.ConfigFile, finalEnv map[string]string) (
 	if err := processEnabled(rendered); err != nil {
 		return nil, err
 	}
+	for ix, service := range rendered.Services {
```
Comment (silvin-lubecki) [inline render/render.go]: I think this feature should be tested (overriding the service images name using the `imageMap`).

**Comment by simonferquel** [comment]
I resolved the CI issue here, @chris-crone @silvin-lubecki PTAL ?

**Comment by simonferquel** [comment]
Rebased + using duffle master branch. @silvin-lubecki PTAL

**Pull request merged**